### PR TITLE
Fix peculiar TFile::GetObject behaviour

### PR DESCRIPTION
**Pull request by dpiparo**
This is a workaround for ROOT-9857

### CODE REVIEW

File: tree/dataframe/test/dataframe_simple.cxx [hunk 1]
```diff
@@ -896,6 +896,28 @@ TEST(RDFSimpleTests, AutomaticNamesOfHisto1DAndGraph)
...
+   TTree *t;
+   f.GetObject(treename, t);
+   RDataFrame df(*t);
+   *df.Define("xy", [](int x, int y) { return x * y; }, {"x", "y"})
+       .Filter([](int xy) { return xy > 0; }, {"xy"})
+       .Count();
+
+   gSystem->Unlink(filename);
+}
+
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: this does not fail/crash in current master

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 2]
```diff
@@ -124,7 +124,22 @@ static std::string GetTreeFullPath(const TTree &tree)
...
+      // threads we trigger ROOT-9857, ending up manipulating the same tree (same pointer)
+      // concurrently.
+      const auto motherDirPath = motherDir->GetPath();
+      const char *fullPathCstr;
+      // Remove from the full path the name of the file if any
+      if (auto currFile = tree.GetCurrentFile()) {
+         const auto currFileName = currFile->GetName();
+         const auto currFileNameLen = strlen(currFileName);
+         if (0 == strncmp(motherDirPath, currFileName, currFileNameLen)) {
+            fullPathCstr = motherDirPath + (currFileNameLen + 1); // +1 to skip the '/'
```
Comment (bluehood) [inline tree/treeplayer/src/TTreeProcessorMT.cxx]: this kind of manipulation might be simpler with `string_view`

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 3]
```diff
@@ -124,7 +124,22 @@ static std::string GetTreeFullPath(const TTree &tree)
...
+      // In case the file name appears in the path of the mother directory
+      // we need to remove it, otherwise, if we open the same file in at least 2
+      // threads we trigger ROOT-9857, ending up manipulating the same tree (same pointer)
+      // concurrently.
+      const auto motherDirPath = motherDir->GetPath();
+      const char *fullPathCstr;
+      // Remove from the full path the name of the file if any
+      if (auto currFile = tree.GetCurrentFile()) {
+         const auto currFileName = currFile->GetName();
+         const auto currFileNameLen = strlen(currFileName);
```
Comment (bluehood) [inline tree/treeplayer/src/TTreeProcessorMT.cxx]: this is `currFile->GetName().Length()`

File: tree/dataframe/test/dataframe_simple.cxx (hunk 1 above)
Comment (dpiparo) [reply tree/dataframe/test/dataframe_simple.cxx]: lol. I am missing the flushing every 2 entries. Coming.

Comment (dpiparo) [reply tree/dataframe/test/dataframe_simple.cxx]: no wait. It does fail

Comment (bluehood) [reply tree/dataframe/test/dataframe_simple.cxx]: ok, never on my machine -- I tried both a debug and a Release build :man_shrugging:

Comment (bluehood) [reply tree/dataframe/test/dataframe_simple.cxx]: I take it back! Was running single-thread :sweat:

Comment (dpiparo) [reply tree/dataframe/test/dataframe_simple.cxx]: I fell in the same trap before starting to flush the input file every 2 events (I had 1 single cluster)...

File: tree/treeplayer/src/TTreeProcessorMT.cxx [hunk 4]
```diff
@@ -124,7 +124,22 @@ static std::string GetTreeFullPath(const TTree &tree)
...
+      // we need to remove it, otherwise, if we open the same file in at least 2
+      // threads we trigger ROOT-9857, ending up manipulating the same tree (same pointer)
+      // concurrently.
+      std::string_view motherDirPath(motherDir->GetPath());
+      std::string_view fullPathCstr;
+      // Remove from the full path the name of the file if any
+      if (auto currFile = tree.GetCurrentFile()) {
+         std::string_view currFileName(currFile->GetName());
+         const auto currFileNameLen = currFileName.size();
+         if (0 == motherDirPath.compare(0, currFileNameLen, currFileName)) {
```
Comment (pcanal) [inline tree/treeplayer/src/TTreeProcessorMT.cxx]: What if the file contains a prefix (root://, file://) ?

**Pull request closed without merging**